core/app/models/spree/promotion_action.rb

```diff
@@ -19,5 +19,15 @@ class PromotionAction < Spree::Base
     def perform(options = {})
       raise 'perform should be implemented in a sub-class of PromotionAction'
     end
+
+    def duplicate
+      duplicated_action = self.dup
+      duplicate_calculator(duplicated_action) if try(:calculator)
```

Comment:
Hrm so the question you want to ask here is "Does the subclass define the calculator method?"

To be a bit more explicit, we might want to say:

```
if responds_to?(:calculator) 
```